src/NuGet.Clients/NuGet.VisualStudio.Implementation/Utility/PathLookupTrie.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace NuGet.VisualStudio
+{
+    /// <summary>
+    /// A data structure providing efficient lookup for known directory containing a file with any nesting level.
+    /// Inspired by prefix string search trie.
+    /// </summary>
+    /// <typeparam name="TValue">A data type stored at leaf nodes. For instance, package ID, GUID, etc.</typeparam>
+    public sealed class PathLookupTrie<TValue>
+    {
+        private static readonly char[] PathSeparators = new[] { Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar };
+
+        private readonly PathSegmentTrieNode _root = new PathSegmentTrieNode(string.Empty);
+
+        /// <summary>
+        /// Indexer setting/retrieving custom data from a leaf node.
+        /// </summary>
+        /// <param name="path">Directory or file path</param>
+        /// <returns>Value associated with leaf node if found.</returns>
+        public TValue this[string path]
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(path))
+                {
+                    throw new ArgumentNullException(nameof(path));
+                }
+
+                var segments = SplitPath(path);
+
+                var current = _root;
+                foreach (var segment in segments)
```

Comment:
a for loop and returning an array from SplitPath could help perf here, this will be called a lot and we've seen issues with this in other areas such as ContentModel where all paths are split